scrapy/resolver.py

```diff
@@ -30,3 +48,60 @@ def getHostByName(self, name, timeout=None):
     def _cache_result(self, result, name):
         dnscache[name] = result
         return result
+
+
+@implementer(IHostnameResolver)
+class CachingHostnameResolver:
+    """
+    Experimental caching resolver. Resolves IPv4 and IPv6 addresses,
+    does not support setting a timeout value for DNS requests.
+    """
+
+    def __init__(self, reactor, cache_size):
+        self.reactor = reactor
+        self.original_resolver = reactor.nameResolver
+        dnscache.limit = cache_size
+
+    @classmethod
+    def from_crawler(cls, crawler, reactor):
+        if crawler.settings.getbool('DNSCACHE_ENABLED'):
+            cache_size = crawler.settings.getint('DNSCACHE_SIZE')
+        else:
+            cache_size = 0
+        return cls(reactor, cache_size)
+
+    def install_on_reactor(self):
+        self.reactor.installNameResolver(self)
+
+    def resolveHostName(self, resolutionReceiver, hostName, portNumber=0,
+                        addressTypes=None, transportSemantics='TCP'):
+
+        @provider(IResolutionReceiver)
+        class CachingResolutionReceiver(resolutionReceiver):
+
+            def resolutionBegan(self, resolution):
+                super(CachingResolutionReceiver, self).resolutionBegan(resolution)
+                self.resolution = resolution
+                self.resolved = False
+
+            def addressResolved(self, address):
+                super(CachingResolutionReceiver, self).addressResolved(address)
+                self.resolved = True
+
+            def resolutionComplete(self):
+                super(CachingResolutionReceiver, self).resolutionComplete()
+                if self.resolved:
+                    dnscache[hostName] = self.resolution
+
+        try:
+            result = dnscache[hostName]
+        except KeyError:
+            result = self.original_resolver.resolveHostName(
+                CachingResolutionReceiver(),
+                hostName,
+                portNumber,
+                addressTypes,
+                transportSemantics
+            )
+        finally:
+            return result
```

Comment:
What would happen if `self.original_resolver.resolveHostName` raises an error? Would it be NameError exception? I think it makes sense to remove `finally` block, and either return directly (`return dnscache[hostName]`), or return result in the end, without finally block.

Reply:
Nice catch, thanks!